common/error.h

```diff
@@ -137,7 +137,7 @@ class ErrorBuilder {
   // Accumulates string message to a temporary `ErrorBuilder`. After streaming,
   // the builder must be converted to an `Error` or `ErrorOr`.
   template <typename T>
-  [[nodiscard]] auto operator<<(const T& message) && -> ErrorBuilder&& {
```

Comment:
I think we should probably keep this one (and add it to the other `<<` overloads for this type: the usual convention is that you can ignore the return value of `<<` if you don't want to stream anymore, so the risk of mistakes is relatively high. Arguably the comment on this function already explains why the attribute is needed.

Reply:
My expectation is that the idea that it could have a mistake is little different from other API calls. That there's a non-nodiscard version is, to me, more indicative of the problem that we're adding it where no bugs result, and so it's more churn than value. Can you give an example of how this would be forgotten in code?

Reply:
I agree this is a closer call, and I don't feel too strongly about it. And yes, the other overload is much more debatable, because it's `&`-qualified, so the return value is presumably available by other means. 

> Can you give an example of how this would be forgotten in code, or a bug that resulted from the overload?

I'm thinking of things like

```c++
ProgramError(source_location) << "Error message";
```

> What makes this different from, for example, a factory function which we would not nodiscard?

Discarded-return-value bugs primarily happen when the programmer doesn't realize that the function has a return value in the first place, or is mistaken about the nature of the return value. Factory functions (and other functions that primarily return a value) are normally named after what they return, which makes that kind of mistake highly unlikely. This operator doesn't have that protection, because the name `<<` (in a streaming rather than bit-shifting context) is normally used for functions whose return value is ignorable.

Reply:
I'm not sure I agree with this reasoning, but as I said, I don't feel too strongly.